### PR TITLE
Finish Card

### PR DESCRIPTION
**Pull request by lauuta185**
### Summary
In this PR, i add the following features:

- News list in tab "News".
- "Refresh" icon when the user makes a pull down.
-  If the user liked a new that appeared liked.

### Card
[https://trello.com/c/5YcsrxPo/24-u4-maquetado-home-news](url)

### Review
Screenshot_20221102_162737

### CODE REVIEW

File: app/src/main/java/com/example/wnews/HomeActivity.kt [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
 import com.google.android.material.tabs.TabLayout
 
+
 class HomeActivity : AppCompatActivity() {
+    lateinit var binding: ActivityHomeBinding
+    lateinit var mainViewModel: NewsViewModel
     override fun onCreate(savedInstanceState: Bundle?) {
+        binding = ActivityHomeBinding.inflate(layoutInflater)
+        val view = binding.root
+        setContentView(view)
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/HomeActivity.kt]: Are you using 'view' again?  Or only in this line? Maybe you can improve it.

File: app/src/main/java/com/example/wnews/HomeActivity.kt [hunk 2]
```diff
@@ -21,8 +37,13 @@ class HomeActivity : AppCompatActivity() {
...
         tabLayout.getTabAt(1)!!.setIcon(R.mipmap.ic_profile_on)
 
+        }
     }
-}
+
+
+
+
+
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/HomeActivity.kt]: Too much spaces 🙈

File: app/src/main/java/com/example/wnews/LoginActivity.kt [hunk 3]
```diff
@@ -39,6 +38,8 @@ class LoginActivity : AppCompatActivity() {
 
                 mainViewModel.validateSecondName(secondNameInput)
 
+
+
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/LoginActivity.kt]: I know pressing enter it's free , but please don't do that

File: app/src/main/java/com/example/wnews/LoginActivity.kt [hunk 4]
```diff
@@ -121,45 +126,87 @@ class LoginActivity : AppCompatActivity() {
...
         }
 
-        fun loadData(): String? {
-            val sharedPreferences =
-                this.getSharedPreferences(SHARED_PREFS, MODE_PRIVATE)
-            return sharedPreferences.getString(KEY_EMAIL, "")
-        }
+        mainViewModel.userId.observe(this) {
+                saveId(it)
+            println(it)
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/LoginActivity.kt]: Delete this

File: app/src/main/java/com/example/wnews/LoginActivity.kt [hunk 5]
```diff
@@ -121,45 +126,87 @@ class LoginActivity : AppCompatActivity() {
...
+
+    fun saveId( id : Int) {
+        val sharedPreferences = this.getSharedPreferences(SHARED_PREFS, MODE_PRIVATE)
+        val editor = sharedPreferences.edit()
+        editor.putInt(USER_ID, id)
+
+        editor.apply()
+    }
+
+    fun loadData(): String? {
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/LoginActivity.kt]: Are you using this?

File: app/src/main/java/com/example/wnews/LoginActivity.kt [hunk 6]
```diff
@@ -121,45 +126,87 @@ class LoginActivity : AppCompatActivity() {
...
+        val sharedPreferences =
+            this.getSharedPreferences(SHARED_PREFS, MODE_PRIVATE)
+        return sharedPreferences.getString(KEY_EMAIL, "")
+    }
+
+    fun loadDataPassword(): String? {
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/LoginActivity.kt]: Same here

File: app/src/main/java/com/example/wnews/NewsAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        val news = mList[position]
+
+        val userLikes = news.likes.contains(UserId)
+
+        if (userLikes) {
+            binding.likeButton.setBackgroundResource(R.mipmap.ic_like_on)
+        }
+
+        Glide.with(binding.newsImage)
+            .load("https://upload.wikimedia.org/wikipedia/commons/thumb/6/65/No-Image-Placeholder.svg/1665px-No-Image-Placeholder.svg.png")
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/NewsAdapter.kt:43]: Move the string values into companion object.

File: app/src/main/java/com/example/wnews/NewsFragment.kt [hunk 8]
```diff
@@ -1,59 +1,68 @@
...
+        inflater: LayoutInflater,
+        container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
-        // Inflate the layout for this fragment
-        return inflater.inflate(R.layout.fragment_news, container, false)
+        mainViewModel = ViewModelProvider(this)[NewsViewModel::class.java]
+        binding = FragmentNewsBinding.inflate(inflater, container, false)
+        val view = binding.root
+        return view
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/NewsFragment.kt:29]: Same question here. Why you put 'binding.root' into a variable that only is used only one time

File: app/src/main/java/com/example/wnews/NewsFragment.kt [hunk 9]
```diff
@@ -1,59 +1,68 @@
...
+        binding.listRecyclerview.layoutManager = LinearLayoutManager(requireContext())
+        mainViewModel.getNews()
+
+        val data = ArrayList<News>()
+        println(data)
+
+        val onCellClickListener = { pressed: News ->
+            mainViewModel.isPressedFav.value = !mainViewModel.isPressedFav.value!!
+           val newTest =mainViewModel.apiNewsResult.value?.data?.find { new: News -> new == pressed }?.commenter
+            println(newTest)
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/NewsFragment.kt]: 👀

File: app/src/main/java/com/example/wnews/viewModel/LoginViewModel.kt [hunk 10]
```diff
@@ -66,9 +66,20 @@ class LoginViewModel (private val app:Application ): AndroidViewModel(app) {
             val result = repository.loginWithUser(email, password)
             apiResult.value = result.isSuccessful
             apiLoading.value = false
+
+
+            val accessToken = result.headers().get(name = "Access-Token").toString()
+            val uid = result.headers().get(name = "Uid").toString()
+            val client = result.headers().get(name = "Client").toString()
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/viewModel/LoginViewModel.kt]: Hardcode string

File: app/src/main/java/com/example/wnews/viewModel/NewsViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        val userId : MutableLiveData<Int> = MutableLiveData()
+        val response = NewsRepository()
+
+        fun getNews() {
+            sharedPreferences =
+                app.getSharedPreferences("sharedPrefs", AppCompatActivity.MODE_PRIVATE)
+            val token = sharedPreferences.getString("Access-Token", "").toString()
+            val uid = sharedPreferences.getString("Uid", "").toString()
+            val client = sharedPreferences.getString("Client", "").toString()
+            val id = sharedPreferences.getInt("id", 0)
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/viewModel/NewsViewModel.kt]: Hardcode string

File: app/src/main/java/com/example/wnews/viewModel/NewsViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            userId.value = id
+            viewModelScope.launch {
+                val apiResult = response.getNews(accessToken = token, uid = uid, client = client)
+
+                if (apiResult.code() == 200) {
+                    val noNewsToView = null
+                    val data = apiResult.body()
+                    apiNewsResult.value = data!!
+                } else if (apiResult.code() == 401) {
+                    println("Unauthorized")
```
Comment (bdecorneliis) [inline app/src/main/java/com/example/wnews/viewModel/NewsViewModel.kt]: 👀

**Review by bdecorneliis — CHANGES_REQUESTED** [review]
We have the PR! We know it was complicated and frustrating but we see that code after code you are getting the idea. Keep going so!  🚀 🚀 🚀 🚀 
I leave some comments to improve.

File: app/src/main/java/com/example/wnews/NewsFragment.kt (hunk 9 above)
Comment (lauuta185) [reply app/src/main/java/com/example/wnews/NewsFragment.kt]: I do not remove this because I will need it later...

**Review by bdecorneliis — APPROVED** [review]
Much better!

**Pull request merged**